src/script/interpreter.cpp

```diff
@@ -1587,29 +1695,538 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    }
+                }
+                break;
+
+                case OP_PUSHCURRENTINPUTINDEX:
+                {
+                    // OP_INSPECTOUTPUT is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    if (checker.GetnIn() > 0x7fffffff)
+                        return set_error(serror, SCRIPT_ERR_UNKNOWN_ERROR);
+                    CScriptNum currIdx = CScriptNum((int64_t) checker.GetnIn());
+                    stack.push_back(currIdx.getvch());
+                }
+                break;
+
+                case OP_INSPECTOUTPUTASSET:
+                case OP_INSPECTOUTPUTVALUE:
+                case OP_INSPECTOUTPUTNONCE:
+                case OP_INSPECTOUTPUTSCRIPTPUBKEY:
+                {
+                    // OP_INSPECTOUTPUT is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    if (stack.size() < 1)
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+
+                    int idx = CScriptNum(stacktop(-1), fRequireMinimal).getint();
                     popstack(stack);
+
+                    auto outs = checker.GetTxvOut();
+                    if (idx < 0 || (unsigned int)idx >= outs->size())
+                        return set_error(serror, SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS);
+                    const CTxOut out = outs->at(idx);
+
+                    switch (opcode)
+                    {
+                        case OP_INSPECTOUTPUTASSET:
+                        {
+                            pushasset(stack, out.nAsset);
+                            break;
+                        }
+                        case OP_INSPECTOUTPUTVALUE:
+                        {
+                            pushvalue(stack, out.nValue);
+                            break;
+                        }
+                        case OP_INSPECTOUTPUTNONCE:
+                        {
+                            valtype vchOutNonce;
+                            if (out.nNonce.IsNull()) {
+                                stack.push_back(vchFalse);
+                            } else {
+                                vchOutNonce.insert(vchOutNonce.begin(), out.nNonce.vchCommitment.begin(), out.nNonce.vchCommitment.begin() + 33);
+                                stack.push_back(vchOutNonce);
+                            }
+                            break;
+                        }
+                        case OP_INSPECTOUTPUTSCRIPTPUBKEY:
+                        {
+                            valtype vchScriptPubKey;
+                            vchScriptPubKey.insert(vchScriptPubKey.begin(), out.scriptPubKey.begin(), out.scriptPubKey.end());
```

Comment:
What happens if/when the max item size is exceeded?

Reply:
Good catch, this should be checked here. I was only considering taproot outputs which are 34 bytes. Nut this introspection allows inspecting other outputs too which could be more than 520. And potentially this can be a DoS vector.

Reply:
We could return the hash of the pubkey, though that requires some caching.
Maybe we should return a parsed pubkey, with the segwit version pushed, and the segwit program pushed, and -1 pushed if it isn't native segwit.

Reply:
I like this idea. We just push the version and program separately. They are also constrained to 40 bytes by BIP141.

Reply:
Since people may want to check their OP_RETURN data outputs too, we probably want both.

I don't know, maybe for non-segwit outputs we should push a hash of the pubkey, and for segwit output we push the parsed version?  Maybe it is okay to just hash in place to start with and optimize with a cache later?

Reply:
In the non-segwit case, How about we fail if `spk > MAX_SCRIPT_ELEMENT_SIZE` and push the script pubkey as it is?

`OP_RETURN`s are restricted by `MAX_OP_RETURN_RELAY = 83` by policy rules, so I don't think we are restricting things severely by doing this.